### PR TITLE
chore: update deps and change edition to 2021

### PR DESCRIPTION
**Pull request by elpiel**
- [x] fix(WIP): core changes - Some events need implementing #40 
- [x] ~chore: updating `jni` requires changes to `jni_ext` module~ #41 
- [x] We need to update `core` after this PR is merged as it reverts a changed field name: https://github.com/Stremio/stremio-core/pull/486
- [x] ~Use StreamingService::base_url in `Stream`~
- [x] Updated Continue Watching model and items
- [x] Add toggle notifications and notifications dismissed events

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -20,20 +20,13 @@ jobs:
...
-      - name: Add i686 target
-        run: rustup target add i686-linux-android
-      - name: Add x86_64 target
-        run: rustup target add x86_64-linux-android
+
+      - uses: dtolnay/rust-toolchain@stable
+        with:
+          components: clippy, rustfmt
+          # Add armv7, aarch64, i686, x86_64 targets
+          targets: armv7-linux-androideabi, aarch64-linux-android, i686-linux-android, x86_64-linux-android
```
Comment (elpiel) [inline .github/workflows/release.yml:28]: Should be tested on the next release.
- [ ] We should also consider running `cargo check` with the given targets to make sure we don't find issues when we want to release a new version

File: Cargo.toml [hunk 2]
```diff
@@ -2,7 +2,10 @@
 name = "stremio-core-android"
 version = "1.1.0"
 authors = ["Smart Code OOD"]
-edition = "2018"
+edition = "2021"
+resolver = "2"
+
+rust-version = "1.66.1"
```
Comment (elpiel) [inline Cargo.toml:8]: We build with stable rust but we keep a MSRV as well.

File: Cargo.toml [hunk 3]
```diff
@@ -13,45 +16,49 @@ lto = true
...
-prost-build = "0.11.9"
-protox = "0.3.3"
+prost-build = "0.11"
+protox = "0.4"
 glob = "0.3.0"
+
+# [patch.'https://github.com/Stremio/stremio-core']
+
+# stremio-core = { path = "../core" }
+# stremio-watched-bitfield = { path = "../core/stremio-watched-bitfield" }
```
Comment (sleeyax) [inline Cargo.toml:64]: Can this be removed or is it some WIP left-overs?

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -3,25 +3,27 @@ name: Build
...
       - name: Checkout
-        uses: actions/checkout@v2
-      - name: Set default rust toolchain
-        run: rustup default 1.61.0
-      - name: Install rustfmt
-        run: rustup component add rustfmt
-      - name: Install clippy
-        run: rustup component add clippy
+        uses: actions/checkout@v3
+      - uses: dtolnay/rust-toolchain@stable
```
[the remark this answers is not in the document]
Comment (elpiel) [reply .github/workflows/build.yml:19]: Usually in the sense that when new rust stable version is released, clippy and fmt checks will start breaking.
It's good to keep the latest stable though and make sure we're complying with the MSRV (it builds) but use `stable`.

File: Cargo.toml (hunk 3 above)
Comment (elpiel) [reply Cargo.toml:64]: This is not a WIP. When developing or updating core it's easier to uncomment these patches and use your path core changes instead for a branch. We've done this for `core-web` as well.

File: src/commonMain/rust/bridge/stream.rs [hunk 5]
```diff
@@ -131,18 +131,32 @@ impl
...
         ),
     ) -> types::Stream {
-        let streaming_server_url =
-            ctx.map(|ctx| ctx.profile.settings.streaming_server_url.to_owned());
-        let deep_links = if stream_request.is_some() && meta_request.is_some() {
-            StreamDeepLinks::from((
+        // in calls that have None for ctx this would panic if we don't set it to default.
+        let settings = ctx
+            .map(|ctx| ctx.profile.settings.to_owned())
+            .unwrap_or_default();
```
Comment (elpiel) [inline src/commonMain/rust/bridge/stream.rs:137]: - [ ] The streaming url should be taken from `StreamingServer::base_url` instead!

File: src/commonMain/rust/bridge/event.rs [hunk 6]
```diff
@@ -18,6 +18,12 @@ impl ToProtobuf<runtime::Event, ()> for Event {
                     runtime::event::LibraryItemsPushedToStorage { ids: ids.clone() },
                 )
             }
+            Event::StreamsPushedToStorage { uid: _ } => {
+                todo!()
+            }
+            Event::NotificationsPushedToStorage { ids: _ } => {
+                todo!()
+            }
```
[the remark this answers is not in the document]
Comment (sleeyax) [reply src/commonMain/rust/bridge/event.rs:26]: See https://github.com/Stremio/stremio-core-kotlin/pull/40 for the implementation

File: src/commonMain/rust/bridge/stream.rs [hunk 7]
```diff
@@ -131,18 +131,32 @@ impl
...
+
+        let deep_links = match (stream_request, meta_request) {
+            (Some(stream_request), Some(meta_request)) => {
+                // when https://github.com/Stremio/stremio-core/pull/490 is merged:
+                // StreamDeepLinks::from((self, *stream_request, *meta_request, &settings))
```
Comment (elpiel) [inline src/commonMain/rust/bridge/stream.rs:142]: - [ ] Should be removed

File: src/commonMain/rust/bridge/stream.rs [hunk 8]
```diff
@@ -131,18 +131,32 @@ impl
...
+                StreamDeepLinks::from((
+                    self,
+                    *stream_request,
+                    *meta_request,
+                    &ctx.map(|ctx| ctx.profile.settings.streaming_server_url.clone()),
+                    &settings,
+                ))
+            }
+            // when https://github.com/Stremio/stremio-core/pull/490 is merged:
+            // _ => StreamDeepLinks::from((self, &settings)),
```
Comment (elpiel) [inline src/commonMain/rust/bridge/stream.rs:152]: - [ ] Should be removed

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -19,21 +19,14 @@ jobs:
...
-        run: rustup target add aarch64-linux-android
-      - name: Add i686 target
-        run: rustup target add i686-linux-android
-      - name: Add x86_64 target
-        run: rustup target add x86_64-linux-android
+        run: yes | sdkmanager "ndk;25.2.9519653"" >/dev/null
+
+      - uses: dtolnay/rust-toolchain@stable
+        with:
+          components: clippy, rustfmt
```
Comment (elpiel) [inline .github/workflows/release.yml:26]: - [ ] remove components, we don't need them.

File: src/commonMain/rust/bridge/stream.rs (hunk 5 above)
Comment (elpiel) [reply src/commonMain/rust/bridge/stream.rs:137]: @TheBeastLT @sleeyax any ideas where/how to best get the (model) `StreamingServer::base_url` and pass it to this function?

Comment (TheBeastLT) [reply src/commonMain/rust/bridge/stream.rs:137]: Why would we need to take it from `StreamingServer`? Streaming server uses the url from settings, that's what was used here. Please revert this.

**Review by TheBeastLT — APPROVED** [review]

File: src/commonMain/rust/bridge/stream.rs (hunk 5 above)
Comment (elpiel) [reply src/commonMain/rust/bridge/stream.rs:137]: Not always. If there's another running server (like starting Stremio service before desktop) you can have 2 different servers.
For consistency with web, we should strive for the same behavior in my opinion and this is where we get the url from for web:


https://github.com/Stremio/stremio-core-web/blob/c6dc920197bc233f4306f99d2fde27023d67d986/src/model/serialize_player.rs#L103-L107

But if you say consistency in this case doesn't matter that much we can leave it as it is. After all on androidtv and android we should see more than 1 instance of the server running, right?

File: build.gradle.kts [hunk 10]
```diff
@@ -69,7 +69,7 @@ kotlin {
 
 android {
     compileSdk = 33
-    ndkVersion = "21.0.6113669"
+    ndkVersion = "25.2.9519653"
```
[the remark this answers is not in the document]
Comment (elpiel) [reply build.gradle.kts:72]: It has been some time that we've looked into this.
There were some problems with the android build and we had to bump up the version.

@sleeyax can also confirm that we've tested this code + indeed we have to bump it some times :)

File: src/commonMain/rust/bridge/stream.rs (hunk 5 above)
Comment (TheBeastLT) [reply src/commonMain/rust/bridge/stream.rs:137]: Web uses streaming server instance, as it needs to check if there's a server running. On android and android tv we don't need that as it depends on the server and expects it to be running.
But even if there are 2 server running, it doesn't matters, as like I said `StreamingServer` will use the url defined in settings, the 2nd running server has no impact.

Comment (elpiel) [reply src/commonMain/rust/bridge/stream.rs:137]: Ok, then we can leave it as it is.
I'm resolving this question as Not applicable.

File: src/commonMain/rust/bridge/continue_watching_preview.rs [hunk 11]
```diff
@@ -0,0 +1,43 @@
+use stremio_core::deep_links::LibraryItemDeepLinks;
```
Comment (TheBeastLT) [inline src/commonMain/rust/bridge/continue_watching_preview.rs:1]: Probably it's better to name this file as `ContinueWatchingItem` to match what it's working with.

**Review by TheBeastLT — APPROVED** [review]

**Review by sleeyax — COMMENTED** [review]
Code looks good but I have trouble building this on my end. Check slack.

**Comment by elpiel** [comment]
Closing in favour of: #48

**Pull request closed without merging**